// shellcheck/build/configs/config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.config = void 0;
const node_url_1 = __importDefault(require("node:url"));
const node_path_1 = __importDefault(require("node:path"));
const env_1 = require("./env");
/**
 * Configuration.
 */
exports.config = {
    bin: `shellcheck${process.platform === 'win32' ? '.exe' : ''}`,
    binDir: node_path_1.default.normalize(`${__dirname}/../../bin`),
    mode: 0o755,
    downloadURL: new node_url_1.default.URL(`https://github.com/vscode-shellcheck/shellcheck-binaries/releases/download`),
    apiURL: new node_url_1.default.URL(`https://api.github.com/repos/vscode-shellcheck/shellcheck-binaries/releases/latest`),
    release: env_1.env.SHELLCHECKJS_RELEASE,
    binaries: {
        linux: {
            platform: 'linux',
            architectures: [
                ['x64', 'x86_64'],
                ['arm64', 'aarch64']
            ]
        },
        darwin: {
            platform: 'darwin',
            architectures: [
                ['x64', 'x86_64'],
                ['arm64', 'aarch64']
            ]
        },
        win32: {
            platform: 'windows',
            architectures: [['x64', 'x86_64']]
        }
    },
    logger: { level: env_1.env.SHELLCHECKJS_LOGGER_LEVEL }
};
